pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/OffloadIndexBlockImpl.java

```diff
@@ -0,0 +1,198 @@
+import org.apache.pulsar.broker.s3offload.OffloadIndexBlock;
+import org.apache.pulsar.broker.s3offload.OffloadIndexEntry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Beta
+public class OffloadIndexBlockImpl implements OffloadIndexBlock {
+    private static final Logger log = LoggerFactory.getLogger(OwnershipCache.class);
+
+    private static final int indexMagicWord = 1000;
+
+    private LedgerMetadata segmentMetadata;
+    private TreeSet<OffloadIndexEntryImpl> indexEntries;
+
+    private final Handle<OffloadIndexBlockImpl> recyclerHandle;
+
+    private static final Recycler<OffloadIndexBlockImpl> RECYCLER = new Recycler<OffloadIndexBlockImpl>() {
+        @Override
+        protected OffloadIndexBlockImpl newObject(Recycler.Handle<OffloadIndexBlockImpl> handle) {
+            return new OffloadIndexBlockImpl(handle);
+        }
+    };
+
+    private OffloadIndexBlockImpl(Handle<OffloadIndexBlockImpl> recyclerHandle) {
+        this.recyclerHandle = recyclerHandle;
+    }
+
+    public static OffloadIndexBlockImpl get(LedgerMetadata metadata, List<OffloadIndexEntryImpl> entries) {
+        OffloadIndexBlockImpl block = RECYCLER.get();
+        block.indexEntries = Sets.newTreeSet();
+        entries.forEach(entry -> block.indexEntries.add(entry));
+        checkState(entries.size() == block.indexEntries.size());
+        block.segmentMetadata = metadata;
+        return block;
+    }
+
+    public static OffloadIndexBlockImpl get(InputStream stream) throws IOException {
+        OffloadIndexBlockImpl block = RECYCLER.get();
+        block.indexEntries = Sets.newTreeSet();
+        block.fromStream(stream);
+        return block;
+    }
+
+    public void recycle() {
+        segmentMetadata = null;
+        indexEntries.clear();
+        indexEntries = null;
+        if (recyclerHandle != null) {
+            recyclerHandle.recycle(this);
+        }
+    }
+
+    @Override
+    public OffloadIndexEntry getEntry(long messageEntryId) {
+        if(messageEntryId > segmentMetadata.getLastEntryId()) {
+            log.warn("Try to get entry: {}, which beyond lastEntryId {}, return null",
+                messageEntryId, segmentMetadata.getLastEntryId());
+            return null;
+        }
+        // find the greatest mapping Id whose entryId <= messageEntryId
+        return this.indexEntries.floor(OffloadIndexEntryImpl.of(messageEntryId, 0, 0));
+    }
+
+    @Override
+    public int getEntryCount() {
+        return this.indexEntries.size();
+    }
+
+    @Override
+    public LedgerMetadata getLedgerMetadata() {
+        return this.segmentMetadata;
+    }
+
+    private static byte[] buildLedgerMetadataFormat(LedgerMetadata metadata) {
+        checkState(metadata instanceof org.apache.bookkeeper.client.LedgerMetadata);
+
+        org.apache.bookkeeper.client.LedgerMetadata ledgerMetadata =
+            (org.apache.bookkeeper.client.LedgerMetadata) metadata;
+
+        return ledgerMetadata.serialize();
```

Comment:
We should avoid using the org.apache.bookkeeper.client.LedgerMetadata serialization. It stores a lot of stuff which isn't necessary for our usecase, and could change under our feet without us noticing. It would be better to define a new protobuf for LedgerMetadata in the pulsar protobuf formats and use that to store only the metadata that is useful for org.apache.bookkeeper.client.api.LedgerMetadata.

Reply:
@ivankelly, Yes, tried to write the serialise/de-serialize by ourselves. The trick part is that we  have to init an instance of LedgerMetadata, If not leverage org.apache.bookkeeper.client.LedgerMetadata, then we may also need to provide an implementation of api.LedgerMetadata here, this seems weird. 
Since client.LedgerMetadata is the only implementation, it may not a too bad thing to use it.